library/bignum_core.h

```diff
@@ -817,4 +821,16 @@ void mbedtls_mpi_core_from_mont_rep(mbedtls_mpi_uint *X,
                                     mbedtls_mpi_uint mm,
                                     mbedtls_mpi_uint *T);
 
+/*
+ * Can't define thread local variables with our abstraction layer: do nothing if threading is on.
+ */
+#if defined(MBEDTLS_TEST_HOOKS) && !defined(MBEDTLS_THREADING_C)
+extern int mbedtls_mpi_optionally_safe_codepath;
```

Comment:
`mbedtls_mpi_optionally_safe_codepath` gives me hives. It's not at all obvious that turning on `MBEDTLS_TEST_HOOKS` doesn't change the functional behavior of the code: I had to carefully check the places where `mbedtls_mpi_optionally_safe_codepath` is used and verify that when the library reads it, it only influences how the library might modify it and nothing else.

`mbedtls_mpi_optionally_safe_codepath` should be documented at the place where it's declared. But the documentation belongs in the test code... The solution to this conundrum is that the variable should be defined in test code. The library code should call a tracing function through a function pointer (doing nothing if the function pointer is null).

This would normally be a blocker. I'm prepared to waive it under time pressure for the 3.6.1 release if we schedule a high-priority follow-up with the assurance that it will not get forgotten on the big tech debt heap.

Reply:
We could declare that the 3.6.1 EPIC can't be closed until that follow-up is done, so that we have to do the follow-up before the end of the quarter.

Reply:
Reopened the followup to address this: https://github.com/Mbed-TLS/mbedtls/pull/9493